### PR TITLE
Import from twittner/quicksink repo.

### PR DESCRIPTION
**Pull request by twittner**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    future: Option<T>,
+    param: Option<S>,
+    state: State,
+    _mark: std::marker::PhantomData<(A, E)>
+}
+
+impl<S, F, T, A, E> SinkImpl<S, F, T, A, E> {
+    /// Project the closure from the pinned struct.
+    fn lambda(self: Pin<&mut Self>) -> &mut F {
+        unsafe {
```
Comment (mxinden) [inline src/lib.rs]: Couldn't you also do this with https://github.com/taiki-e/pin-project without `unsafe`?

Comment (twittner) [reply src/lib.rs]: Ok, but I went with https://github.com/taiki-e/pin-project-lite to minimise dependencies.

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+                Action::Send(x) => stdout.write_all(x).await?,
+                Action::Flush => stdout.flush().await?,
+                Action::Close => stdout.close().await?
+            }
+            Ok::<_, io::Error>(stdout)
+        });
+
+        task::block_on(async move {
+            let values = vec![Ok(&b"hello\n"[..]), Ok(&b"world\n"[..])];
+            assert!(stream::iter(values).forward(sink).await.is_ok())
```
Comment (mxinden) [inline src/lib.rs:253]: Wouldn't it make more sense to use something other than `io::stdout` to check whether `["hello", "world"]` comes out on the other end?

**Review by mxinden — APPROVED** [review]
Implementation looks good to me.

File: src/lib.rs (hunk 2 above)
Comment (twittner) [reply src/lib.rs:253]: Absolutely.

**Review by mxinden — APPROVED** [review]
Thanks for the improved test!

**Pull request merged**